content/en/blog/_posts/2024-07-31-consistent-read-from-cache.md

```diff
@@ -0,0 +1,83 @@
+---
+layout: blog
+title: 'Kubernetes Consistent Reads from Cache: A Leap in Performance and Scalability'
+date: 2024-07-31
+slug: consistent-read-from-cache-beta
+author: >
+  Marek Siarkowicz (Google)
+---
+
+Kubernetes is renowned for its robust orchestration of containerized applications,
+but as clusters grow, the demands on the control plane can become a bottleneck.
+A key challenge has been ensuring strongly consistent reads from the etcd datastore,
+requiring resource-intensive quorum reads.
+
+Today, the Kubernetes community is excited to announce a major improvement:
+**Consistent Reads from Cache**, graduating to Beta in Kubernetes 1.31.
+
+### Why consistent reads matter
+
+Consistent reads are essential for ensuring that Kubernetes components,
+like the kubelet, have an accurate view of the latest cluster state.
+Guaranteeing consistent reads is crucial for maintaining the accuracy and reliability of Kubernetes operations,
+enabling components to make informed decisions based on up-to-date information.
+In large-scale clusters, fetching and processing this data can be a performance bottleneck,
+especially for operations like frequent pod listings by kubelets.
```

Comment:
Kubelets are not doing frequent listing.

I suggest to switch to talk more broadly about "many component instances listing large number of objects" or sth like that.

Reply:
Can rewrite, but this comes from the original KEP:

https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md#motivation

```
We expect the biggest gain to be from node-originating requests (e.g. kubelet listing pods scheduled on its node). For those requests, the size of the response is small (it fits a single page, assuming you won't make it extremely small), whereas the number of objects to process is proportional to cluster-size (so fairly big). For example, when kubelets requests pods schedule against it in a 5k node cluster with 30pods/node, the kube-apiserver must list the 150k pods from etcd and then filter that list down to the list of 30 pods that the kubelet actually need. This must occur for each list request from each of the 5k kubelets. If served from watch cache, this same request can be served from built-in index filtering out the 30 pods each kubelet needs from the data in the cache.
```

Reply:
I know, but that's not true - so let's fix it.

Reply:
I thought a little about this and remembered the optimistic case we were testing at the begging, the problem with apiserver object filtering. I updated the text to say that kubelet benefits from this because it makes filtered requests for just one node, but to serve this request we needed to load all pods from etcd. Is this more accurate representation?

Reply:
I'm fine with the current text.